### PR TITLE
Feature/add page user profile editing

### PR DESCRIPTION
**Pull request by alls77**

### CODE REVIEW

File: backend/Perflow/Common/Security/SecurityHelper.cs [hunk 1]
```diff
@@ -17,6 +18,22 @@ public static class SecurityHelper
...
-            );
+            );
+
+        public static bool ValidatePassword(string password, string hash, string salt)
+        {
+            return HashPassword(password, Convert.FromBase64String(salt)) == hash;
+        }
+
+        public static byte[] GetRandomBytes(int length = 32)
+        {
```
Comment (proskurnicha) [inline backend/Perflow/Common/Security/SecurityHelper.cs:29]: Magic value should be in const

File: frontend/src/app/modules/shared/directives/calendar/semantic-calendar.directive.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        type: this.calendarType,
+        monthFirst: false,
+        formatter: {
+          date: (date:Date) => {
+            if (!date) return '';
+            const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate();
+            const month = date.getMonth() + 1 < 10 ? `0${date.getMonth() + 1}` : date.getMonth() + 1;
+            const year = date.getFullYear();
+            return `${year}-${month}-${day}`;
+          }
```
Comment (proskurnicha) [inline frontend/src/app/modules/shared/directives/calendar/semantic-calendar.directive.ts]: The magic value (10) should be in const

File: frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+@Component({
+  selector: 'app-profile-edit-form',
+  templateUrl: './profile-edit-form.component.html',
+  styleUrls: ['./profile-edit-form.component.sass']
+})
+export class ProfileEditFormComponent implements OnInit {
+  form!: FormGroup;
+  genders: { key: boolean; text: string; }[] = [{ key: false, text: 'Male' }, { key: true, text: 'Female' }];
+  countries: Country[] = countries;
+  standartIcon: string = '../../../../assets/images/standartIcon.png';
```
Comment (proskurnicha) [inline frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts]: Is it possible to use a semantic icon? (or font-awesome)

File: frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+@Component({
+  selector: 'app-profile-edit-form',
+  templateUrl: './profile-edit-form.component.html',
+  styleUrls: ['./profile-edit-form.component.sass']
+})
+export class ProfileEditFormComponent implements OnInit {
+  form!: FormGroup;
+  genders: { key: boolean; text: string; }[] = [{ key: false, text: 'Male' }, { key: true, text: 'Female' }];
```
Comment (proskurnicha) [inline frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts]: Add interface  Gender {
 key: boolean;
text: string;
}

File: frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  get userName() {
+    return this.form.get('userName')!;
+  }
+
+  get description() {
+    return this.form.get('description')!;
+  }
+
+  onSelect(event: Event) {
+    const file = (<HTMLInputElement>event.target).files![0];
```
Comment (proskurnicha) [inline frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts]: Could we use some angular function here?

File: frontend/src/app/modules/user/update-password-form/update-password-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    this.form = new FormGroup({
+      currentPassword: new FormControl('', Validators.required),
+      newPassword: new FormControl('', [
+        Validators.required,
+        Validators.pattern('^(?=.*[A-Za-z])(?=.*[0-9@$!%*#?&-_])[A-Za-z0-9@$!%*#?&-_]{6,20}$')
+      ]),
+      newPasswordConfirmation: new FormControl('', [
+        Validators.required,
+        Validators.pattern('^(?=.*[A-Za-z])(?=.*[0-9@$!%*#?&-_])[A-Za-z0-9@$!%*#?&-_]{6,20}$')
+      ])
```
Comment (proskurnicha) [inline frontend/src/app/modules/user/update-password-form/update-password-form.component.ts]: Use const for pattern. Copy paste here
^(?=.*[A-Za-z])(?=.*[0-9@$!%*#?&-_])[A-Za-z0-9@$!%*#?&-_]{6,20}$

File: frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    return this.form.get('email')!;
+  }
+
+  get userName() {
+    return this.form.get('userName')!;
+  }
+
+  get description() {
+    return this.form.get('description')!;
+  }
```
Comment (proskurnicha) [inline frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts:46]: There is a way to get rid of these getters and setters, use ngModel. Read about it here.
In this code, we have only 3 fields(getter), but imagine if we have 20 fields so we should have 20 getters.

https://angular.io/guide/forms

File: frontend/src/app/modules/user/update-password-form/update-password-form.component.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  public reset() {
+    this.form.reset();
+  }
+
+  public onSubmit() {
+    const updatedUserPassword: UserChangePassword = {
+      currentPassword: this.form.get('currentPassword')!.value,
+      newPassword: this.form.get('newPassword')!.value,
+      id: this.user.id
+    };
```
Comment (proskurnicha) [inline frontend/src/app/modules/user/update-password-form/update-password-form.component.ts]: The same here.
There is a way to get rid of these getters, use ngModel. Read about it here.

https://angular.io/guide/forms

**Review by proskurnicha — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts (hunk 5 above)
Comment (alls77) [reply frontend/src/app/modules/user/profile-edit-form/profile-edit-form.component.ts]: I didn't find angular function for it https://blog.angular-university.io/angular-file-upload/

File: frontend/src/app/modules/shared/directives/message/semantic-message.directive.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+@Directive({
+  selector: '[appSmMessage]'
+})
+
+export class SemanticMessageDirective implements AfterViewInit {
+  constructor(private _message: ElementRef) { }
+
+  ngAfterViewInit() {
+    const element = $(this._message.nativeElement);
+    element.on('click', () => element.closest('.message').transition('fade'));
```
Comment (NikitaKrasnovSW) [inline frontend/src/app/modules/shared/directives/message/semantic-message.directive.ts]: Replace with onClick?

**Review by NikitaKrasnovSW — CHANGES_REQUESTED** [review]

**Review by proskurnicha — APPROVED** [review]

**Comment by proskurnicha** [comment]
Please fix the conflicts, the rest looks good for me. Nice job!

**Comment by alls77** [comment]
conflicts fixed

**Pull request merged**